src/main/java/org/opensearch/ad/ml/MLMetricsCorrelationInputBuilder.java

```diff
@@ -0,0 +1,230 @@
+
+        for (String metricKey : metricKeys) {
+            List<AnomalyResult> metricAnomalies = metricGroups.get(metricKey);
+            List<Double> timeSeries = buildTimeSeries(metricAnomalies, executionStartTime, allBucketList, DEFAULT_BUCKET_SIZE_MILLIS);
+            matrix.add(timeSeries);
+        }
+
+        log.info("Built correlation matrix: {} metrics x {} time buckets", matrix.size(), matrix.isEmpty() ? 0 : matrix.get(0).size());
+
+        return new MLMetricsCorrelationInput(
+            matrix,
+            metricKeys,
+            detectorMetadataMap,
+            executionStartTime,
+            executionEndTime,
+            DEFAULT_BUCKET_SIZE_MILLIS,
+            bucketTimestamps
+        );
+    }
+
+    private static Map<String, List<AnomalyResult>> groupByMetric(List<AnomalyResult> anomalies) {
+        Map<String, List<AnomalyResult>> groups = new HashMap<>();
+
+        for (AnomalyResult anomaly : anomalies) {
+            String metricKey = getMetricKey(anomaly);
+            groups.computeIfAbsent(metricKey, k -> new ArrayList<>()).add(anomaly);
+        }
+
+        return groups;
+    }
+
+    private static String getMetricKey(AnomalyResult anomaly) {
+        String detectorId = anomaly.getDetectorId();
+        Optional<Entity> optEntity = anomaly.getEntity();
+        if (!optEntity.isPresent()) {
+            // single stream detector
+            return detectorId;
+        }
+
+        // hc detector
+        Entity entity = optEntity.get();
+        String entityKey = entity
+            .getAttributes()
+            .entrySet()
+            .stream()
+            .map(e -> e.getKey() + "=" + e.getValue())
+            .sorted()
+            .collect(Collectors.joining(","));
+
+        return detectorId + "|" + entityKey;
+    }
+
+    /**
+     * Build time series for a single metric.
+     * Returns array of anomaly scores for each bucket (0.0 if no anomaly in that bucket).
+     */
+    private static List<Double> buildTimeSeries(
+        List<AnomalyResult> metricAnomalies,
+        Instant executionStartTime,
+        List<Integer> allBucketList,
+        long bucketSizeMillis
+    ) {
+        // Build map of bucket_index -> max_anomaly_score
+        Map<Integer, Double> bucketScores = new TreeMap<>();
+        long startTimeMillis = executionStartTime.toEpochMilli();
+
+        for (AnomalyResult anomaly : metricAnomalies) {
+            long anomalyTime = anomaly.getDataStartTime().toEpochMilli();
+            int bucketIndex = (int) ((anomalyTime - startTimeMillis) / bucketSizeMillis);
+
+            if (bucketIndex >= 0) {
+                // Use MAX score if multiple anomalies in same bucket
+                double currentScore = bucketScores.getOrDefault(bucketIndex, 0.0);
+                double newScore = anomaly.getAnomalyScore();
+                bucketScores.put(bucketIndex, Math.max(currentScore, newScore));
```

Comment:
Should we consider interval? Our anomalies are interval anomalies. We can put anomalies scores to all of the buckets interleaving current interval [data start, data end].

Reply:
synced offline, no changes needed here